sparse_strips/vello_common/src/glyph.rs

```diff
@@ -623,3 +656,331 @@ mod tests {
+    fn maintain(&mut self) {
+        // Maximum number of full renders where we'll retain an unused glyph
+        const MAX_ENTRY_AGE: u32 = 64;
+        // Maximum number of full renders before we force a prune
+        const PRUNE_FREQUENCY: u32 = 64;
+        // Always prune if the cached count is greater than this value
+        const CACHED_COUNT_THRESHOLD: usize = 256;
+        // Number of encoding buffers we'll keep on the free list
+        const MAX_FREE_LIST_SIZE: usize = 128;
+
+        let free_list = &mut self.free_list;
+        let serial = self.serial;
+        self.serial += 1;
+        // Don't iterate over the whole cache every frame
+        if serial - self.last_prune_serial < PRUNE_FREQUENCY
+            && self.cached_count < CACHED_COUNT_THRESHOLD
+        {
+            return;
+        }
+        self.last_prune_serial = serial;
+        self.static_map.retain(|_, entry| {
+            if serial - entry.serial > MAX_ENTRY_AGE {
+                if free_list.len() < MAX_FREE_LIST_SIZE {
+                    free_list.push(core::mem::take(&mut entry.path));
+                }
+                self.cached_count -= 1;
+                false
+            } else {
+                true
+            }
+        });
+        self.variable_map.retain(|_, map| {
+            map.retain(|_, entry| {
+                if serial - entry.serial > MAX_ENTRY_AGE {
+                    if free_list.len() < MAX_FREE_LIST_SIZE {
+                        free_list.push(core::mem::take(&mut entry.path));
+                    }
+                    self.cached_count -= 1;
+                    false
+                } else {
+                    true
+                }
+            });
+            !map.is_empty()
+        });
+    }
+
+    fn clear(&mut self) {
+        self.free_list.clear();
+        self.static_map.clear();
+        self.variable_map.clear();
+        self.cached_count = 0;
+        self.serial = 0;
+        self.last_prune_serial = 0;
+    }
+}
+
+struct OutlineCacheSession<'a> {
+    map: &'a mut HashMap<OutlineKey, OutlineEntry>,
+    free_list: &'a mut Vec<OutlinePath>,
+    serial: u32,
+    cached_count: &'a mut usize,
+}
+
+impl<'a> OutlineCacheSession<'a> {
+    fn new(outline_cache: &'a mut OutlineCache, var_key: VarLookupKey<'_>) -> Self {
+        let map = if var_key.0.is_empty() {
+            &mut outline_cache.static_map
+        } else {
+            // This is still ugly in rust. Choices are:
+            // 1. multiple lookups in the hashmap (implemented here)
+            // 2. always allocate and copy the key
+            // 3. use unsafe
+            // Pick 1 bad option :(
+            if outline_cache.variable_map.contains_key(&var_key) {
+                outline_cache.variable_map.get_mut(&var_key).unwrap()
+            } else {
+                outline_cache
+                    .variable_map
+                    .entry(var_key.into())
+                    .or_default()
+            }
```

Comment:
Would this approach work?

```rust
match outline_cache
  variable_map
  .raw_entry_mut()
  .from_key(&var_key)
{
    RawEntryMut::Occupied(entry) => entry.into_mut(),
    RawEntryMut::Vacant(entry) => entry.insert(var_key.into(), HashMap::new()).1,
}
```

Reply:
You're a genius @grebmeg . Thank you. This is a huge improvement. Implemented in https://github.com/linebender/vello/pull/1215/commits/8a7b5150475e4c679f1bf7002ec2dbe28bb0961d